delta/plugins/storage/src/main/scala/ch/epfl/bluebrain/nexus/delta/plugins/storage/storages/model/StorageValue.scala

```diff
@@ -1,20 +1,24 @@
 package ch.epfl.bluebrain.nexus.delta.plugins.storage.storages.model
 
 import akka.http.scaladsl.model.Uri
+import cats.implicits._
+import ch.epfl.bluebrain.nexus.delta.plugins.storage.storages.model.EncryptionState._
 import ch.epfl.bluebrain.nexus.delta.sdk.implicits._
+import ch.epfl.bluebrain.nexus.delta.plugins.storage.storages.model.StorageRejection.InvalidEncryptionSecrets
 import ch.epfl.bluebrain.nexus.delta.rdf.jsonld.context.JsonLdContext.keywords
 import ch.epfl.bluebrain.nexus.delta.sdk.model.Label
-import ch.epfl.bluebrain.nexus.delta.sdk.model.identities.AuthToken
 import ch.epfl.bluebrain.nexus.delta.sdk.model.permissions.Permission
-import io.circe.Encoder
 import io.circe.generic.extras.Configuration
 import io.circe.generic.extras.semiauto.deriveConfiguredEncoder
 import io.circe.syntax._
+import io.circe.Encoder
 
 import java.nio.file.Path
 import scala.annotation.nowarn
 
-sealed trait StorageValue extends Product with Serializable {
+sealed trait StorageValue[A <: EncryptionState] extends Product with Serializable {
```

Comment:
I tried going back and forth between the two.

However the other approach generates too much code and repetition while not having an obvious advantage. 
The current way works well since `Storage` is force to have a `StorageValue[Decrypted]` and `StorageEvent` is forced to have a `StorageValue[Encrypted]`.

In terms of serialization, each serializer should take care of what it wants to do with the values. In this case `StorageEvent` serializer takes care of its serialization and `Storage` takes care of its own.

Reply:
The code and the tests makes sure that having a `StorageValue[Decrypted]` already provides the correctly decrypted fields, so there is no reason to have a concrete type for those fields.

Same goes for `StorageValue[Encrypted]`

Reply:
The fact that secrets are encoded as String is a problem because they can be displayed on the logs by our code but also akka code and others and we should do wrap them anyway.
If we just want encryption at rest, it should not generate too much code as it will be handled by the encoder/decoder.
Where would it be verbose ?

Reply:
I don't understand the problem with the logs. The concern with the logs or akka displaying credentials can be easily fixed by overriding the `toString`

Reply:
Absolutely but my point is that we should do it with StorageFields, StorageConfig where this can be solved by wrapping the sensitive strings.
We could also use this wrapper to properly encode/decode the token/secrets

Reply:
I get it

Reply:
I've fixed that adding a `Secret` case class and used it everywhere where there could be sensitive data involved (fields, config, source)